File: beam_fdm.m
% Parameters
L = 1;   % Length of the domain
T = 10;  % Total time
M = 10;   % Number of spatial points
N = 10000;  % Number of time steps

% Other special parameters
beta = 0.1;
obstaclePosition = -2;
kappa = 0.01;

% Velocity and RHS
initialVelocity = 1;
force = -10;
t0 = - obstaclePosition / initialVelocity;

alpha = 0.0001;

% Tolerance
Tol = 1e-3;

% Spatial and time steps
dx = L / M;
dt = (T - t0) / N;

% Initialize solution arrays
u = zeros(N,M);

% Initial condition
u(1,:) = 0;  
u(2,:) = initialVelocity * dt;

% Time-stepping loop
for j = 2:N-1
    for i = 3:M-2
        u(j+1,i) = 2 * u(j,i) ...
                   - u(j-1,i) ...
                   + alpha^2 * (dt^2 / dx^4) * (u(j,i+2) - 4 * u(j,i+1) + 6 * u(j,i) - 4 * u(j,i-1) + u(j,i-2)) ...
                   + (dt^2) * force;
    end
    
    % Left end boundary condition
    u(j+1,1) = u(j+1,2);

    % Contact conditions at the right end
    if abs(u(j,M) - obstaclePosition) < Tol || u(j,M) < obstaclePosition

        u(j+1,M-1) = - ((dt * alpha^2) / (2 * dx^2 * beta)) * (2 * u(j,M-2) - u(j,M-3) - u(j,M)) ...
                     - ((dt * kappa) / beta) * (u(j,M-1) - obstaclePosition) ...
                     + u(j-1,M-1);

        u(j+1,M) = u(j+1,M-1);
    else
        u(j+1,M-1) = u(j+1, M-2);
        u(j+1,M) = u(j+1, M-1);
    end
end

disp(u);

% Plot results
x_values = linspace(0, L, M);  
figure;
plot(x_values, u(N,:), 'r', 'DisplayName', 'Left end');        
yline(obstaclePosition, '--g', 'DisplayName', 'Obstacle');  % Obstacle
xlabel('Time (t)');
ylabel('u');
title('FDM Solution');
legend show;
hold off;

File: beam_fdm.py
import numpy as np
import math
import matplotlib.pyplot as plt

# Parameters
L = 1   # Length of the domain
T = 10   # Total time
M = 100  # Number of spatial points
N = 10000  # Number of time steps
c = 2  # Wave speed

# Other special parameters
beta = 0.1
h = 1
k = 0.1

#Velocity and RHS
v0 = 1
f = 10
t0 = -h/v0

#Tolerance
Tol = 1e-12



# Spatial and time steps
dx = L / M
dt = (T-t0) / N

# phi
phi = 1/(1 + k*dx + beta*dx/dt)

# Initialize solution arrays
u = np.zeros((N+1, M+1))


# Initial condition
u[0,:] = 0  
u[1,:] = v0 * dt


for n in range(1, N):
    for i in range(1, M):
        u[n+1,i] = 2*(1 - c**2 * (dt/dx**2)**2) * u[n,i] + \
                    c**2 * (dt/dx**2)**2 * (u[n,i+2] - 4*u[n,i+1]  - 4*u[n,i-1] + u[n,i-2]) - \
                    u[n-1,i] + \
                    (dt**2)*f
                    
    #left End
    u[n+1,0] = u[n+1,1]

    #Contact conditions at right end
    if abs(u[n,M] - h) < Tol or u[n,M] > h: 
        u[n+1,M] = phi * (u[n+1,M-1] + k*h*dx + beta*(dx/dt)*u[n,M])
 
    else:
        u[n+1,M] = u[n+1, M - 1]


# Plot results
plt.plot(u[:, 0] - L, np.linspace(t0, T, N+1), color = 'red', label = 'u(-l,t)') #left boundary
plt.plot(u[:, -1], np.linspace(t0, T, N+1), color = 'blue', label = 'u(0,t)') #right boundary
plt.axvline(x = h, color = 'g', ls = "--", label = 'h') # obstacle at h 
plt.xlabel('x')
plt.ylabel('u(-l, t) and u(0,t)')
plt.title("FDM")
plt.legend()
plt.show()